### ISSUE TITLE
Trouble deploying on Debian 8.0 linux container

### ISSUE THREAD

**Issue by lhorace** [jprjr/multistreamer#15]
Hi,

I am having a bit of problem deploying multistreamer here. Unfortunately, on your readme, it doesn't indicate which lua version you need to run this program?

**Correction:** The readme does mention the program requires lua 5.1 further down.

In Debian 8.0 (jessie), there are 3 different version available: 5.0, 5.1, 5.2, 5.3.

5.1
```
/usr/bin/lua: error loading module 'posix.ctype' from file '/usr/local/src/multistreamer/lua_modules/lib/lua/5.1/posix.so':
        /usr/local/src/multistreamer/lua_modules/lib/lua/5.1/posix.so: undefined symbol: luaL_fileresult
stack traceback:
        [C]: ?
        [C]: in function 'require'
        ...ltistreamer/lua_modules/share/lua/5.1/posix/init.lua:29: in main chunk
        [C]: in function 'require'
        /usr/local/src/multistreamer/bin/multistreamer.lua:12: in main chunk
        [C]: ?

```

5.2
```
/usr/bin/lua: /usr/local/src/multistreamer/bin/multistreamer.lua:12: module 'posix' not found:
        no field package.preload['posix']
        no file '/usr/local/src/multistreamer/lua_modules/share/lua/5.2/posix.lua'
        no file '/usr/local/src/multistreamer/lua_modules/share/lua/5.2/posix/init.lua'
        no file './posix.lua'
        no file '/usr/local/src/multistreamer/lua_modules/lib/lua/5.2/posix.so'
stack traceback:
        [C]: in function 'require'
        /usr/local/src/multistreamer/bin/multistreamer.lua:12: in main chunk
        [C]: in ?
```

**NOTE 1:** I've never worked with lua in the past.

**NOTE 2:** I assumed I had to use luarocks from /opt/openresty/bin/luarocks because I wasn't aware luarocks was already in the repositories.

**NOTE 3:** LuaRocks from Debian 8.0 repository wants '--tree=<path>'

**NOTE 4:** LuaRocks from Debian 8.0 doesn't install dependencies AFAIK. It just errors out: "Unable to statisfy dependecies". Might be a bug in packaging. See full commands in order below:
```
luarocks --tree=lua_modules install bit32
luarocks --tree=lua_modules install lua-cjson
luarocks --tree=lua_modules install date
luarocks --tree=lua_modules install luacrypto
luarocks --tree=lua_modules install ansicolors
luarocks --tree=lua_modules install lpeg
luarocks --tree=lua_modules install etlua
luarocks --tree=lua_modules install loadkit
luarocks --tree=lua_modules install luafilesystem
luarocks --tree=lua_modules install mimetypes
luarocks --tree=lua_modules install luasocket
luarocks --tree=lua_modules install luabitop
luarocks --tree=lua_modules install pgmoon
luarocks --tree=lua_modules install netstring
luarocks --tree=lua_modules install lua-resty-exec
luarocks --tree=lua_modules install lua-resty-jit-uuid
luarocks --tree=lua_modules install lua-resty-string
luarocks --tree=lua_modules install lua-resty-http
luarocks --tree=lua_modules install lua-resty-upload
luarocks --tree=lua_modules install lapis
luarocks --tree=lua_modules install etlua
luarocks --tree=lua_modules install luaposix
luarocks --tree=lua_modules install luafilesystem
luarocks --tree=lua_modules install whereami
```

I opted for https://github.com/jprjr/htpasswd-auth-server, see requiste luarocks below:

**NOTE:** I am not sure if I should open a seperate issue for htpasswd-auth-server
```
luarocks --tree=lua_modules install lecho
luarocks --tree=lua_modules install bit32
luarocks --tree=lua_modules install luaposix
luarocks --tree=lua_modules install etlua
luarocks --tree=lua_modules install luafilesystem
luarocks --tree=lua_modules install lbase64
```

**Note 1:** htpasswd-auth-server failed to run here, see https://github.com/jprjr/htpasswd-auth-server/issues/1

**Note 2:** In hindsight, htpasswd-auth-server seems like a overkill and I went with nginx instead. Although creating a username and password combo for htpasswd was useful with htpasswd-auth-server.

Your sample systemd unit service has a syntax error:

`systemd[1]: [/etc/systemd/system/multistreamer.service:2] Unknown lvalue 'description' in section 'Unit'`

Description needs to start with a capital 'D'. Additionally, multistreamer errors out with:

`Error: auth_endpoint not set in config`
 
With the following command:

`./bin/multistreamer -e prod run`

However, this seem to work:

`./bin/multistreamer -e production run`

What is the real difference between development and production?

Lastly, is this significant:

```
Jun 25 16:45:57 rtmp multistreamer[4228]: 2017/06/25 16:45:57 [error] 4244#0: stream [lua] init_worker_by_lua:10: running irc thread, context: ngx.timer
Jun 25 16:45:57 rtmp multistreamer[4228]: 2017/06/25 16:45:57 [error] 4243#0: stream [lua] init_worker_by_lua:10: running irc thread, context: ngx.timer
```

PS. Took me 3 to 4 hours to set this up, before I even check if it can actually stream. :D

**EDIT:** It's running, I created a 'Stream', extracted the Stream URL and Stream Key and attempted to stream with OBS.

![deepinscreenshot20170625140603](https://user-images.githubusercontent.com/12678456/27518505-2b6803de-59a7-11e7-94ea-752473f8d9d7.png)

Wireshark TCP dump:

![deepinscreenshot20170625140720](https://user-images.githubusercontent.com/12678456/27518521-6223ff2c-59a7-11e7-8439-89c86091a992.png)

Where can I view the logs in regards to why it's failing for rtmp? There is nothing userful in my /var/log/syslog minus the regular HTTP and lua traffic?

**EDIT:** My config.lua:

```
local config = require('lapis.config').config

config({'development','production'}, {
  -- name of the cookie used to store session data
  session_name = 'multistreamer',

  -- key for encrypting session data
  secret = '*************',

  -- whether to log queries and requests
  logging = {
      queries = true,
      requests = true,
  },

  -- if deploying somewhere other than the root of a domain,
  -- set this to your prefix (ie, '/multistreamer')
  http_prefix = '',

  -- set an rtmp prefix
  -- note: this can only be a single string,
  -- no slashes etc
  -- defaults to 'multistreamer' if unset
  rtmp_prefix = 'multistreamer',

  -- path to your nginx+lua+rtmp binary
  nginx = '/opt/openresty-rtmp/bin/openresty',

  -- path to psql
  psql = '/usr/bin/psql',

  -- path to ffmpeg
  ffmpeg = '/usr/bin/ffmpeg',

  -- set your logging level
  log_level = 'debug',

  -- setup your external urls (without prefixes)
public_http_url = 'http://rtmp.lhprojects.int',
  public_rtmp_url = 'rtmp://rtmp.lhprojects.int:1935',

  -- setup your private (loopback) urls (without prefixes)
  private_http_url = 'http://127.0.0.1:8081',
  private_rtmp_url = 'rtmp://127.0.0.1:1935',

  -- setup your public IRC hostname, for the web
  -- interface
  public_irc_hostname = 'example.com',
  -- setup your public IRC port, to report in the
  -- web interface
  public_irc_port = '7000',
  -- set to true if you've setup an SSL terminator in front
  -- of multistreamer
  public_irc_ssl = true,


  -- configure streaming networks/services
  -- you'll need to register a new app with each
  -- service and insert keys/ids in here

  -- 'rtmp' just stores RTMP urls and has no config,
  networks = {
    -- twitch = {
    --   client_id = 'client_id',
    --   client_secret = 'client_secret',
    --   ingest_server = 'rtmp://somewhere', -- see https://bashtech.net/twitch/ingest.php
                                             -- for a list of endpoints
    -- },
    -- facebook = {
    --   app_id = 'app_id',
    --   app_secret = 'app_secret',
    -- },
    -- youtube = {
    --   client_id = 'client_id',
    --   client_secret = 'client_secret',
    --   country = 'us', -- 2-character country code, used for listing available categories
    -- },
    rtmp = true,
  },

  -- postgres connection settings
  postgres = {
    host = '127.0.0.1',
    user = 'multistreamer',
    password = 'multistreamer',
    database = 'multistreamer'
  },

  -- nginx http "listen" directive, see
  -- http://nginx.org/en/docs/http/ngx_http_core_module.html#listen
  http_listen = '10.0.0.12:8081',

  -- nginx rtmp "listen" directive, see
  -- https://github.com/arut/nginx-rtmp-module/wiki/Directives#listen
  rtmp_listen = '1935',

  -- nginx irc "listen" directive, see
  -- https://nginx.org/en/docs/stream/ngx_stream_core_module.html#listen
  irc_listen = '6667',

  -- set the IRC hostname reported by the server
  irc_hostname = 'rtmp.lhprojects.int',

  -- should users be automatically brought into chat rooms when
  -- their streams go live? (default false)
  -- this is handy for clients like Adium, Pidgin, etc that don't
  -- have a great IRC interface
  irc_force_join = false,

  -- number of worker processes
  worker_processes = 1,

  -- http auth endpoint
  -- multistreamer will make an HTTP request with the 'Authorization'
  -- header to this URL when a user logs in
  -- see http://nginx.org/en/docs/http/ngx_http_auth_request_module.html
  -- see https://github.com/jprjr/ldap-auth-server for an LDAP implementation
  auth_endpoint = 'http://127.0.0.1:8080/',

  -- redis host
  redis_host = '127.0.0.1:6379',

  -- prefix for redis keys
  redis_prefix = 'multistreamer/',

  -- path to trusted ssl certificate store
  ssl_trusted_certificate = '/etc/ssl/certs/ca-certificates.crt',

  -- dns resolver
  dns_resolver = '8.8.8.8',

  -- maximum ssl verify depth
  ssl_verify_depth = 5,

  -- sizes for shared dictionaries (see https://github.com/openresty/lua-nginx-module#lua_shared_dict)
  lua_shared_dict_streams_size = '10m',
  lua_shared_dict_writers_size = '10m',

  -- specify the run directory to hold temp files etc,
  -- defaults to $HOME/.multistreamer if not set
  -- work_dir = '/path/to/some/folder',

  -- set the path to sockexec's socket
  -- see https://github.com/jprjr/sockexec for installation details
  sockexec_path = '/tmp/exec.sock',

})
```

**Comment by jprjr** [comment]
Re: luarocks/lua - I usually install luarocks compiled against the luajit included with OpenResty, something like:

```
./configure \
  --prefix=/opt/openresty-rtmp \
  --with-lua=/opt/openresty-rtmp/luajit \
  --lua-suffix=jit \
  --with-lua-include=/opt/openresty-rtmp/luajit/include/luajit-2.1
```

The newer versions of OpenResty/LuaJIT have a hard time compiling the lua-cjson module, so I'll compile it against Lua 5.1. LuaJIT is compatible with Lua 5.1 modules.

That's good to know about the Debian version of luarocks not installing dependencies automatically, I've only really used the newer versions of luarocks which handle dependencies. I'll get the README updated with the full list of modules, and change it to use that `--tree=lua_modules` syntax, that way its compatible with newer and older luarocks.

The reason `-e prod` failed - the `config.lua` has a line like:

```
config({'development','production'}, {
```

And you need that parameter to match one of those environment names (the `lapis` docs have more info on keeping multiple environment's configs together). On my box, I use `dev` and `prod` and didn't think about how my names don't mass the example config. I've updated the systemd file to fix that `description` problem, and to use `-e production`.

Regarding checking logs, they should be under /var/log/syslog, systemd should automatically direct them there.

Another way to see why it won't go live: go into the Add/Remove accounts screen and change "require preview before going live" to "Yes"

Start pushing video, go into the Edit Metadata screen, and click "go live" - if there's an error, starting the video push, you'll see the error message via web interface. I can look into saving error messages so they're viewable in the web interface when "require preview" is "no."

**Comment by jprjr** [comment]
Also, I'd double-check your DNS, I get two responses for that domain, only one seems to respond to ping etc:

```
dig rtmp.lhprojects.int

; <<>> DiG 9.10.3-P4-Debian <<>> rtmp.lhprojects.int
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 19743
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;rtmp.lhprojects.int.           IN      A

;; ANSWER SECTION:
rtmp.lhprojects.int.    10      IN      A       198.105.244.130
rtmp.lhprojects.int.    10      IN      A       104.239.207.44

;; Query time: 38 msec
;; SERVER: 192.168.1.254#53(192.168.1.254)
;; WHEN: Sun Jun 25 13:53:31 CDT 2017
;; MSG SIZE  rcvd: 80
```

So it could be OBS is connecting to the wrong IP?

**Comment by lhorace** [comment]
Here's an update:

I almost gave up! I was digging through the filesystem to _try_ to figure out why it was failing to connect with multistreamer; then it occur to me that multistreamer was running as it's own user and had it's own home directory /var/lib/multistreamer. Unfortunately it was a dead end or was it? I took a peek at nginx_production.conf to see if I could find anything useful and I did.

```
rtmp {
  server {
    listen 1935;
    chunk_size 4096;
    application multistreamer {
      live on;
      record off;
      meta copy;
      idle_streams off;

      on_publish http://127.0.0.1:8081/on-publish;
      on_publish_done http://127.0.0.1:8081/on-done;
      on_update http://10.0.0.12:8081/on-update;

      hls on;
      hls_path /var/lib/multistreamer/.multistreamer/hls;
      hls_nested on;
      hls_cleanup on;
      hls_continuous off;
      hls_fragment 10s;
      hls_playlist_length 60s;

    }
  }
}
```

Notice the Nginx parameters below:
```
on_publish http://127.0.0.1:8081/on-publish;
on_publish_done http://127.0.0.1:8081/on-done;
on_update http://10.0.0.12:8081/on-update;
```

That refer to:
```
private_http_url = 'http://127.0.0.1:8081',
```

Well, nothing was listening on `127.0.0.1:8081`, and once I change `private_http_url` to `http://10.0.0.12:8081`. Viola, it worked! I tested streaming to Youtube, see https://www.youtube.com/watch?v=kpIiIY4PRmk&feature=youtu.be. The framedrop is because I am streaming 1080p60FPS using x264 and I am barely being able to multi-task here hehe.

Which begs my next question, what is the purpose of public and private url?

I was checking my hypervisor box CPU usage and saw multistreamer (ffmpeg) using about 1.2% of CPU and almost paniced because I misinterpred the graph. Assuming that it was using an entire core (DUMB ME).

> Re: luarocks/lua - I usually install luarocks compiled against the luajit included with OpenResty, something like:

It seems that luarocks from OpenResty have compatibility issue with Debian lua5.1. See first error.

> EDIT: and I'm even more dumb, those DNS responses were from my router which tries to be helpful

I have multistreamer running on a internal box on a internal domain lol. So yeah you wouldn't be able to reach it.

**Comment by jprjr** [comment]
> Which begs my next question, what is the purpose of public and private url?

The idea there is your public URL might be something that has to traverse a WAF or something like that. I wanted to make sure that internal notification calls don't unnecessarily leave the box. On my setup, I have multistreamer listening on 127.0.0.1 and have another instance of nginx listening on my public IPs, providing SSL, etc.

Re: keys - they're needed for dealing with stream metadata, pulling comments, and so on. If you use the custom RTMP URLs, you only push video (and nothing else!). So that's why your chat widget wasn't working.

The default URL for the chat widget only shows publicly-visible YouTube/Facebook//etc comments - so if you're using the "custom rtmp" type there's nothing to show. If you remove "&hide_irc" from the widget URL, it'll show IRC messages, which is a handy way to do a quick "is this working *at all*" kind of test.

For transcoding, you wouldn't use the FFMPEG Pull Args (on the add/remove accounts screen), you'd want to use the FFMPEG Args on the "edit metadata" screen, or on the screen for changing the custom RTMP account settings.

Here's the FFPEG args I use for transcoding to Periscope:

```
-c:v libx264 -x264opts keyint=60 -preset veryfast -b:v 900k -vf scale=960:540 -c:a copy
```

Basically:

* `-c:v libx264` - encode to libx264
* `-x264opts keyint=60` - setting my keyframe interval to 2s (60 frames at 30fps). This is pretty dependent on your stream's settings
* `-preset veryfast` - x264 has encoding presets, I forget them all - but they're the same presets you see in OBS if you enable advanced settings
* `-b:v 900k` - set the video bitrate to 900kbps
* `-vf scale=960:540` - use a scale filter to 960x540 (periscope's preferred resolution)
* `-c:a copy` - copy the audio as-is

The FFMPEG Pull Args are used when you have other stream you want to pull from, so you could use something like "-i http://some-website.com/stream.flv" if you have some other stream you want to pull into multistreamer.

**Comment by lhorace** [comment]
Thanks for the info!

> keys - they're needed for dealing with stream metadata, pulling comments, and so on. If you use the custom RTMP URLs, you only push video (and nothing else!).

My problem with this, like I had with Restream, is that I like to create my events which Restream was using Stream Now to stream. How does that work with Multistreamer? Can I specifically pick an event?

Re: FFMPEG: I will be streaming later on and I'll evaluate how transcoding performs.

**Comment by jprjr** [comment]
Multistreamer uses events, but it doesn't try to read in any events - when you stream, it creates a new event on-the-fly.

On the "edit metadata" screen, you can setup the basic event details - like title, tags, category, and description. So then you go live, it will automatically fill in those details, so you can still get everything setup ahead of time like you do with events.

The overall idea is to present "Live Now"-type functionality to the user, but for multiple accounts on a single page. That was the reason I started writing this, actually - I was dual-streaming to Twitch and YouTube using RTMP, and I hated having to pull up multiple websites for updating the stream metadata.

I think I could implement an event-picker, though. I could just have a drop-down that lists any upcoming events, and include a "Create new" option, so picking "create new" would just keep the existing functionality, and picking an event would make Multistreamer ignore whatever you set for the title, description, etc, and just push video to that event.

**Comment by jprjr** [comment]
Hey there! I've got a new release out (9.1.0) that supports YouTube events. It's listed in a drop-down on the dashboard, if you choose an event all the other metadata settings are ignored, it will just use whatever you've already put into the event.

**Comment by lhorace** [comment]
Sorry about being MIA, I haven't live stream for quite awhile. I was going to deploy this on my OpenVZ CentOS boxed hosted on cloud provider until I read the title I created with this ticket. Damn! 

I'll try to give a shot and offer my findings and use this ticket as a reference to work through installation issues! :-(
